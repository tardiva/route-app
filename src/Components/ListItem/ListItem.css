.list-item {
    margin: 3px 0;
    padding: 0 15px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    height: 50px;
    width: 100%;
    text-align: left;
    background-color: #fff;
    border: 1px solid #BEBEBE;
    border-radius: 4px;
}

.list-item__content {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
}

.list-item__delete-button {
    padding: 0;
    border: none;
    outline: none;
    background: none;
    color: #282c34;
    cursor: pointer;
}