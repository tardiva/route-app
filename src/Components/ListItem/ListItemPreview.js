import React, {Component} from 'react';
import Icon from 'react-icons-kit';
import {ic_close} from 'react-icons-kit/md/ic_close';

class ListItemPreview extends Component {

  render() {
    return (
      <div className="list-item" style={this.props.style}>
        <div className="list-item__content">{this.props.name}</div>
        <button className="list-item__delete-button">
          <Icon icon={ic_close}/>
        </button>
      </div>)
  }
}

export default ListItemPreview;