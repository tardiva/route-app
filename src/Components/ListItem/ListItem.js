import React, { Component } from 'react';
import { findDOMNode } from 'react-dom'
import { DragSource, DropTarget } from 'react-dnd';
import './ListItem.css';
import Icon from 'react-icons-kit';
import {ic_close} from 'react-icons-kit/md/ic_close';

const itemSource = {
  beginDrag(props, monitor, component) {

    const offsetWidth = (findDOMNode(component)).offsetWidth;

    return {
      id: props.item.id,
      index: props.index,
      name: props.item.name,
      width: offsetWidth
    }
  },
};

const itemTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return
    }

    // Determine rectangle on screen
    const hoverBoundingRect = (findDOMNode(component)).getBoundingClientRect();

    // Get vertical middle
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the top
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return
    }

    // Time to actually perform the action
    props.moveItem(dragIndex, hoverIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    monitor.getItem().index = hoverIndex
  }
};

class ListItem extends Component {

  constructor(props) {
    super(props);
    this.handleRemove = this.handleRemove.bind(this);
  }

  render() {
    const {
      item,
      isDragging,
      connectDragSource,
      connectDropTarget,
    } = this.props;
    const opacity = isDragging ? 0 : 1;
    return (
      connectDragSource &&
      connectDropTarget &&
      connectDragSource(
        connectDropTarget(
          <li className="list-item" style={{opacity}}>
            <div className="list-item__content">{item.name}</div>
            <button className="list-item__delete-button" onClick={this.handleRemove}>
              <Icon icon={ic_close}/>
            </button>
          </li>)
      )
    );
  }

  handleRemove() {
    this.props.removeItem(this.props.index);
  }
}

export default DropTarget('listItem', itemTarget, (connect) => ({
  connectDropTarget: connect.dropTarget(),
}))(DragSource('listItem', itemSource, (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging(),
}))(ListItem));