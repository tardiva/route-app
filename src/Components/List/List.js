import React, { Component } from 'react';
import './List.css';
import { DragDropContext } from 'react-dnd';
import MultiBackend, { Preview } from 'react-dnd-multi-backend';
import HTML5toTouch from 'react-dnd-multi-backend/lib/HTML5toTouch';
import ListItem from '../ListItem/ListItem';
import ListItemPreview from '../ListItem/ListItemPreview'

export class List extends Component {

  render() {
    return (
      <div className="list">
        <ul className="list__list">
          {this.props.items.map((item, i) => {
            return (<ListItem key={item.id}
                              index={i}
                              item={item}
                              moveItem={this.props.moveItem}
                              removeItem={this.props.removeItem}
            />)
          })}
        </ul>
        <Preview generator={this.generatePreview} />
      </div>
    );
  }

  generatePreview(type, item, style) {
    Object.assign(style, {width: `${item.width}px`});
    return (<ListItemPreview name={item.name} style={style} />);
  }
}

export default DragDropContext(MultiBackend(HTML5toTouch))(List);