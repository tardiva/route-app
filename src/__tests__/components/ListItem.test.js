import React, { Component } from 'react';
import { shallow, mount } from 'enzyme';
import ListItem from '../../Components/ListItem/ListItem';
import {List} from '../../Components/List/List';
import points from '../../Fixtures/points';
import TestBackend from 'react-dnd-test-backend';
import { DragDropContext } from 'react-dnd';

// Obtain the reference to the component before React DnD wrapping
const OriginalListItem = ListItem.DecoratedComponent;

// Stub the React DnD connector functions with an identity function
const identity = el => el;

// Wraps a component into a DragDropContext that uses the TestBackend.
function wrapInTestContext(DecoratedComponent) {
   return DragDropContext(TestBackend)(
     class TestContextContainer extends Component {
       render() {
         return <DecoratedComponent {...this.props} />;
       }
     }
   );
 }

it('should render List Item correctly', () => {
  const wrapper = shallow(<OriginalListItem item={points[0]}
                                            connectDragSource={identity}
                                            connectDropTarget={identity}/>);
  expect(wrapper).toMatchSnapshot();
 });

it('should show point name as List Item text', () => {
  const wrapper = shallow(<OriginalListItem item={points[0]}
                                            connectDragSource={identity}
                                            connectDropTarget={identity}/>);
  expect(wrapper.find('.list-item__content').text()).toBe(points[0].name);
 });

it('should call removeItem prop when clicking Remove button', () => {
  const onClickSpy = jest.fn();
  const index = 0;
  const wrapper = shallow(<OriginalListItem index={index}
                                            item={points[index]}
                                            connectDragSource={identity}
                                            connectDropTarget={identity}
                                            removeItem={onClickSpy}/>);
  wrapper.find('.list-item__delete-button').simulate('click');
  expect(onClickSpy).toHaveBeenLastCalledWith(0);
});


describe('test drag and drop list items', () => {

  function testDNDItems(sourceIndex, targetIndex, offsetY, isReordered) {
    const onMoveSpy = jest.fn();

    const TestContext = wrapInTestContext(List);

    const root = mount(<TestContext items={points} moveItem={onMoveSpy}/>);
    const manager = root.instance().getManager();
    const backend = manager.getBackend();
    const monitor = manager.getMonitor();
    jest.spyOn(monitor, 'getClientOffset').mockImplementation(() => { return {y: offsetY} });

    const sourceComponent = root.find('DropTarget(DragSource(ListItem))').at(sourceIndex).find('DragSource(ListItem)').instance();
    const targetComponent = root.find('DropTarget(DragSource(ListItem))').at(targetIndex).instance();
    backend.simulateBeginDrag([sourceComponent.getHandlerId()]);
    backend.simulateHover([targetComponent.getHandlerId()]);

    if (isReordered){
      expect(onMoveSpy).toHaveBeenLastCalledWith(sourceIndex, targetIndex);
    } else {
      expect(onMoveSpy).not.toHaveBeenCalled();
    }
  }

  beforeEach(() => {
    jest.spyOn(Element.prototype, 'getBoundingClientRect').mockImplementation(() => { return { top: 40, bottom: 80 } });
  });

  it('should call moveItem prop when reordering List Items', () => {
    testDNDItems(0, 1, 70, true);
  });

  it('should not call moveItem prop when dragging downwards and the cursor is below less than 50%', () => {
    testDNDItems(0, 1, 45, false);
  });

  it('should not call moveItem prop when dragging upwards and the cursor is above less than 50%', () => {
    testDNDItems(1, 0, 70, false);
  });

  it('should not replace items with themselves', () => {
    testDNDItems(1, 1, 0, false);
  });
});


