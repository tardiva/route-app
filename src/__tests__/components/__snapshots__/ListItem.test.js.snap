// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render List Item correctly 1`] = `
<li
  className="list-item"
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <div
    className="list-item__content"
  >
    My first point
  </div>
  <button
    className="list-item__delete-button"
    onClick={[Function]}
  >
    <Icon
      fill="currentColor"
      icon={
        Object {
          "children": Array [
            Object {
              "attribs": Object {
                "d": "M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z",
              },
              "name": "path",
            },
          ],
          "viewBox": "0 0 24 24",
        }
      }
      size={16}
    />
  </button>
</li>
`;
